Change guide to resolve public addresses using Load Balancer service (#10)
hazelcast-guides/kubernetes-external-client: doc/modules/ROOT/pages/index.adoc

= Connect To Hazelcast Running on Kubernetes from Outside

This is a complete example presenting how to use Hazelcast cluster deployed on Kubernetes with Hazelcast Client running outside of Kubernetes.

include::{templates-url}/link-to-repo.adoc[]

== What You’ll Learn

In this guide you will learn how to deploy Hazelcast Kubernetes cluster and connect to it using a client outside Kubernetes.

== Prerequisites

* Up and running https://kubernetes.io/[Kubernetes] cluster (https://minikube.sigs.k8s.io/docs/[Minikube] is good enough)
* Kubernetes command line tool, https://kubernetes.io/docs/tasks/tools/install-kubectl/[kubectl]

== Configure Hazelcast cluster on Kubernetes

There are two approaches you may want to use when setting up on-premises Hazelcast cluster in Kubernetes:

* **Unisocket Client** - client sends requests to a random Hazelcast member

> * **Smart Client** - client connects to all members and sends requests directly to the right members

Reviewer: nit: "right members" doesn't give a hint to the reader
```suggestion
* **Smart Client** - client connects to all members and sends requests directly to the members owning the data
```

[tabs]
====

Kubectl::
+

--
[source, bash]
----
kubectl apply -f https://raw.githubusercontent.com/hazelcast/hazelcast-kubernetes/master/rbac.yaml

kubectl run hz-hazelcast-0 --image=hazelcast/hazelcast:$HAZELCAST_VERSION -l "role=hazelcast"
kubectl run hz-hazelcast-1 --image=hazelcast/hazelcast:$HAZELCAST_VERSION -l "role=hazelcast"
kubectl run hz-hazelcast-2 --image=hazelcast/hazelcast:$HAZELCAST_VERSION -l "role=hazelcast"

kubectl create service loadbalancer hz-hazelcast --tcp=5701 -o yaml --dry-run=client | kubectl set selector --local -f - "role=hazelcast" -o yaml | kubectl create -f -
----
--

Helm::
+
--
[source, bash]
----
helm repo add hazelcast https://hazelcast-charts.s3.amazonaws.com/
helm repo update
helm install hz-hazelcast --set service.type=LoadBalancer hazelcast/hazelcast
----
--

====

==== Verifying Hazelcast cluster

You can check that the Hazelcast cluster is up and running.

[source, bash]
----
kubectl get pods
NAME             READY   STATUS    RESTARTS   AGE
hz-hazelcast-0   1/1     Running   0          32s
hz-hazelcast-1   1/1     Running   0          30s
hz-hazelcast-2   1/1     Running   0          29s
----

> You can also check that all Hazelcast members formed one cluster.

Reviewer: ```suggestion
You can also check that all Hazelcast members formed a cluster.
```

[source, bash]
----
kubectl logs hz-hazelcast-0
...
Members {size:3, ver:3} [
        Member [10.216.6.7]:5701 - 6d2100e0-8dcf-4e7c-ab40-8e98e23475e3 this
        Member [10.216.5.6]:5701 - 5ab4d554-fd7d-4929-8475-0ddf79a21076
        Member [10.216.8.6]:5701 - 7f7dd5f4-e732-4575-89d6-a6e823da38da
]
----

At this point you have a Hazelcast cluster exposed with a single LoadBalancer service called `hz-hazelcast`. You can find it's address with the following command.

[source, bash]
----
kubectl get service hz-hazelcast
NAME           TYPE           CLUSTER-IP       EXTERNAL-IP       PORT(S)          AGE
hz-hazelcast   LoadBalancer   10.108.141.178   10.96.184.178     5701:31434/TCP   5m44s
----

[NOTE]
====
If you are using Minikube, you need to execute `minikube tunnel` now in order to get LoadBalancer External IPs assigned.
====

> The field `EXERNAL-IP` is the address of your Hazelcast cluster.

Reviewer: ```suggestion
The field `EXTERNAL-IP` is the address of your Hazelcast cluster.
```

=== Connecting with Hazelcast Client

Configure the Hazelcast client with the external address and disable smart routing to use the unisocket connection.

[tabs]
====

Java::
+
--
[source, java]
----
ClientConfig config = new ClientConfig();
config.getNetworkConfig().addAddress("<EXTERNAL-IP>")
                         .setSmartRouting(false);
----
--

NodeJS::
+
--
[source, javascript]
----
const { Client } = require('hazelcast-client');

const clientConfig = {
    network: {
        clusterMembers: [
            '<EXTERNAL-IP>'
        ],
        smartRouting: false
    }
};
const client = await Client.newHazelcastClient(clientConfig);
----
--

Go::
+
--
[source, go]
----
import (
	"log"

	"github.com/hazelcast/hazelcast-go-client"
)

func main() {
    config := hazelcast.NewConfig()
	config.ClusterConfig.SetAddress("<EXTERNAL-IP>:5701")
    config.ClusterConfig.SmartRouting = false
	client, err := hazelcast.StartNewClientWithConfig(config)
	if err != nil {
		log.Fatal(err)

> 	}

Reviewer 2: Somehow these lines have different indentation.

Just an idea, but I like to store `config.ClusterConfig` in a variable if I need to set a few things on it, like:
```go
cc := &config.ClusterConfig
cc.SetAddress("<EXTERNAL-IP>:5701")
cc.SmartRouting = false
```

}
----
--

====

> Finally, start the client application using the following command.

Reviewer: nit: it would be nice to include a sample output or, at least, comment what the client is supposed to do and what the user should see in the logs.

cd go-unisocket
go get github.com/hazelcast/hazelcast-go-client@v1.0.0-preview.4
go run main.go
----
--

====

=== Clean Up

To clean up the environment execute the following commands.

[tabs]
====

Kubectl::
+

--
[source, bash]
----
kubectl delete pod/hz-hazelcast-0 pod/hz-hazelcast-1 pod/hz-hazelcast-2 service/hz-hazelcast
kubectl delete -f https://raw.githubusercontent.com/hazelcast/hazelcast-kubernetes/master/rbac.yaml
----
--

Helm::
+
--
[source, bash]
----
helm uninstall hz-hazelcast
----
====

Unisocket is very simple to configure; however, it has one significant drawback - low performance. Using a load balancer is perfect for traffic distribution across stateless services; however, Hazelcast is not stateless. Actually, Hazelcast is more like a sharded database in which each member contains a different part of the data. That is why it’s suboptimal to randomly load balance the traffic. It effectively means that each operation needs to be internally migrated, because your data may be load balanced to member 1, while the partition for its key is stored in member 2. All in all, if performance is important for your use case, then you need to use smart client.

== Smart Client

> Hazelcast smart client stores the mapping from key to member, which means that it sends the data directly to the member which contains the right data partition. This fact implies that in the Kubernetes environment, we need to provide access to each Hazelcast pod from the outside. A dynamic approach to this problem is to expose each pod with a separate service. Again, the service can be either `LoadBalancer` or `NodePort`. In a real-life scenario, that would probably be `NodePort`, since having a separate public IP for each pod is expensive.

Reviewer: nit: the "client stores the mapping from key to member" phrase may be confusing for the reader
```suggestion
Hazelcast smart client is capable of mapping the given key with its owner member, which means that it sends the data directly to the member which contains the right data partition. This fact implies that in the Kubernetes environment, we need to provide access to each Hazelcast pod from the outside. A dynamic approach to this problem is to expose each pod with a separate service. Again, the service can be either `LoadBalancer` or `NodePort`. In a real-life scenario, that would probably be `NodePort`, since having a separate public IP for each pod is expensive.
```

hz-hazelcast-0   1/1     Running   0          32s
hz-hazelcast-1   1/1     Running   0          30s
hz-hazelcast-2   1/1     Running   0          29s
----

At this point, you should also have every Hazelcast member exposed with a separate externally accessible address.

[source, bash]
----
kubectl get service
NAME             TYPE           CLUSTER-IP       EXTERNAL-IP      PORT(S)          AGE
hz-hazelcast     LoadBalancer   10.219.246.40    35.230.84.127    5701:30443/TCP   4m2s
hz-hazelcast-0   LoadBalancer   10.219.255.141   34.145.108.167   5701:30091/TCP   4m7s
hz-hazelcast-1   LoadBalancer   10.219.241.203   34.82.71.106     5701:30687/TCP   4m5s
hz-hazelcast-2   LoadBalancer   10.219.247.106   35.247.93.190    5701:32452/TCP   4m4s
----

We'll use the `hz-hazelcast` service for the discovery.

[source, bash]
----
kubectl get service hz-hazelcast
NAME           TYPE           CLUSTER-IP      EXTERNAL-IP     PORT(S)          AGE
hz-hazelcast   LoadBalancer   10.219.246.40   35.230.84.127   5701:30443/TCP   5m29s
----

[NOTE]
====
If you are using Minikube, you need to execute `minikube tunnel` now in order to get LoadBalancer External IPs assigned.
====

> The field `EXERNAL-IP` is the address of your Hazelcast cluster.

Reviewer: ```suggestion
The field `EXTERNAL-IP` is the address of your Hazelcast cluster.
```

=== Connecting with Hazelcast Client

Configure the Hazelcast client to connect to the cluster external address.

[tabs]
====

Java::
+
--
[source, java]
----
ClientConfig config = new ClientConfig();
config.getNetworkConfig().addAddress("<EXTERNAL-IP>");
----
--

NodeJS::
+
--
[source, javascript]
----
const { Client } = require('hazelcast-client');

const clientConfig = {
    network: {
        clusterMembers: [

>             '<EXERNAL-IP>'

Reviewer: ```suggestion
            '<EXTERNAL-IP>'
```

        ]
    }
};
const client = await Client.newHazelcastClient(clientConfig);
----
--

Go::
+
--
[source, go]
----
import (
	"log"

	"github.com/hazelcast/hazelcast-go-client"
)

func main() {
    config := hazelcast.NewConfig()
	config.ClusterConfig.SetAddress("<EXTERNAL-IP>:5701")
	client, err := hazelcast.StartNewClientWithConfig(config)
	if err != nil {
		log.Fatal(err)

> 	}

Reviewer 2: The indentation is different from the rest on the first line.

}
----
--
====

> Finally, start the client application using the following command.

Reviewer: nit: the same consideration on the expected client behavior and output as in the unisocket section.

[tabs]
====

Java::
+
[source, bash]
----
cd java
mvn package
java -jar target/*jar-with-dependencies*.jar
----
--

NodeJS::
+
--
[source, bash]
----
cd nodejs
npm install
npm start
----
--

Go::
+
--
[source, bash]
----
cd go

> go get github.com/hazelcast/hazelcast-go-client@v1.0.0-preview.4

Reviewer 2: `go get` step shouldn't be necessary.
